File: src/main/java/info/idgst/digest/DefaultDigestService.java
package info.idgst.digest;

import info.idgst.exception.DigestAlreadyExistsException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

/**
 * Default implementation of {@link DigestService}.
 *
 * @author Aliaksei Bahdanau
 */
@Service
public class DefaultDigestService implements DigestService {

    private final Logger logger = LoggerFactory.getLogger(this.getClass());

    private final DigestCache digestCache;
    private final DigestRepository digestRepository;

    @Autowired
    public DefaultDigestService(final DigestCache digestCache, final DigestRepository digestRepository) {
        this.digestCache = digestCache;
        this.digestRepository = digestRepository;
    }

    @Override
    public Digest findByTitle(String digestTitle) {
        return digestCache.fetchByTitle(digestTitle);
    }

    @Override
    public Digest findById(String digestID) {
        return digestCache.fetch(digestID);
    }

    @Override
    public Digest save(Digest digest) {
        Digest existingDigest = digestCache.fetchByTitle(digest.getTitle());
        if (existingDigest != null) {
            String msg =
                    "Attempt of saving duplicated digest with title: " + digest.getTitle() + "Existing Digest ID: " +
                    existingDigest.getId();
            logger.warn(msg);
            throw new DigestAlreadyExistsException(msg);
        }
        return digestRepository.save(digest);
    }

    @Override
    public Page<Digest> findAll(int page, int size, Sort.Direction sortDirection, String sortBy) {
        return digestCache.fetch(page, size, sortDirection, sortBy);
    }
}

File: src/main/kotlin/info/devbug/digest/service/DefaultDigestService.kt
package info.devbug.digest.service

import info.devbug.digest.DigestCache
import info.devbug.digest.repository.DigestDto
import info.devbug.digest.util.DigestParser
import info.devbug.digest.repository.DigestRepository
import info.devbug.digest.util.DigestReader
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cache.annotation.Cacheable
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service

/**
 * Default implementation of [DigestService].
 *
 * @author Aliaksei Bahdanau.
 */
@Service
open class DefaultDigestService : DigestService{

    private val digestRepository: DigestRepository
    private val digestReader: DigestReader
    private val digestCache: DigestCache

    @Autowired constructor(digestRepository: DigestRepository, digestReader: DigestReader, digestCache: DigestCache) {
        this.digestRepository = digestRepository
        this.digestReader = digestReader
        this.digestCache = digestCache
    }

    /**
     * Returns the [DigestDto] object according to specified the digest file path.
     */
    fun readDigestFromFile(filePath: String): DigestDto {
        return digestReader.readDigest(filePath)
    }

    /**
     * Parses [DigestDto] title and returns current digest number.
     */
    fun getDigestNumber(digestTitle: String): String {
        val digestParser = DigestParser()
        return digestParser.getDigestNumber(digestTitle)
    }

    override fun findAll(page: Int, size: Int): Page<DigestDto> {
        return digestCache.fetch(page, size)
    }

    override fun save(digest: DigestDto): DigestDto {
        val existingDigest = digestCache.fetchByTitle(digest.title)

        if (existingDigest != null && existingDigest.title == digest.title) throw Exception("Already exists")
        return digestRepository.save(digest)
    }

    @Cacheable("searches")
    override fun findById(id: String): DigestDto? {
        return digestCache.fetch(id)
    }
}